File: csharp/src/EcfClient/Comprobantes/Tipos31_34.cs
using EcfClient.Comprobantes.Datos;

namespace EcfClient.Comprobantes;

/// <summary>Factura de Crédito Fiscal (tipo 31).</summary>
public class FacturaCreditoFiscal31 : ComprobanteBase
{
    public override int TipoEcf => 31;

    public double MontoGravadoTotal { get; set; }
    public double MontoGravadoI1 { get; set; }
    public int Itbis1 { get; set; } = 18;
    public double TotalItbis { get; set; }
    public double TotalItbis1 { get; set; }
    public double MontoTotal { get; set; }
    protected override double GetMontoTotalVal() => MontoTotal;

    protected override Dictionary<string, object?> BuildIdDoc()
    {
        var d = base.BuildIdDoc();
        d["FechaVencimientoSecuencia"] = Trim(FechaVencimientoSecuencia);
        d["IndicadorMontoGravado"] = IndicadorMontoGravado.ToString();
        return d;
    }

    protected override Dictionary<string, object?> BuildTotales() => new()
    {
        ["MontoGravadoTotal"] = MontoGravadoTotal.ToString("F2"),
        ["MontoGravadoI1"] = MontoGravadoI1.ToString("F2"),
        ["ITBIS1"] = Itbis1.ToString(),
        ["TotalITBIS"] = TotalItbis.ToString("F2"),
        ["TotalITBIS1"] = TotalItbis1.ToString("F2"),
        ["MontoTotal"] = MontoTotal.ToString("F2")
    };
}

/// <summary>Factura de Consumo (tipo 32). RFCE si MontoTotal &lt; 250_000.</summary>
public class FacturaConsumo32 : ComprobanteBase
{
    public override int TipoEcf => 32;

    public double MontoGravadoTotal { get; set; }
    public double MontoGravadoI1 { get; set; }
    public int Itbis1 { get; set; } = 18;
    public double TotalItbis { get; set; }
    public double TotalItbis1 { get; set; }
    public double MontoTotal { get; set; }
    protected override double GetMontoTotalVal() => MontoTotal;

    protected override Dictionary<string, object?> BuildIdDoc()
    {
        var d = base.BuildIdDoc();
        if (MontoTotal >= 250_000)
            d["FechaVencimientoSecuencia"] = Trim(FechaVencimientoSecuencia);
        d["IndicadorMontoGravado"] = IndicadorMontoGravado.ToString();
        return d;
    }

    protected override Dictionary<string, object?> BuildTotales() => new()
    {
        ["MontoGravadoTotal"] = MontoGravadoTotal.ToString("F2"),
        ["MontoGravadoI1"] = MontoGravadoI1.ToString("F2"),
        ["ITBIS1"] = Itbis1.ToString(),
        ["TotalITBIS"] = TotalItbis.ToString("F2"),
        ["TotalITBIS1"] = TotalItbis1.ToString("F2"),
        ["MontoTotal"] = MontoTotal.ToString("F2")
    };
}

/// <summary>Nota de Débito (tipo 33). Requiere InformacionReferencia.</summary>
public class NotaDebito33 : ComprobanteBase
{
    public override int TipoEcf => 33;
    private readonly InformacionReferencia _infoRef = new();

    public double MontoExento { get; set; }
    public double MontoTotal { get; set; }

    public InformacionReferencia InformacionReferencia => _infoRef;

    protected override Dictionary<string, object?> BuildIdDoc()
    {
        var d = base.BuildIdDoc();
        d["FechaVencimientoSecuencia"] = Trim(FechaVencimientoSecuencia);
        return d;
    }

    protected override Dictionary<string, object?> BuildTotales() => new()
    {
        ["MontoExento"] = MontoExento.ToString("F2"),
        ["MontoTotal"] = MontoTotal.ToString("F2")
    };

    protected override InformacionReferencia? BuildInformacionReferencia() => _infoRef;
}

/// <summary>Nota de Crédito (tipo 34). Requiere InformacionReferencia.</summary>
public class NotaCredito34 : ComprobanteBase
{
    public override int TipoEcf => 34;
    private readonly InformacionReferencia _infoRef = new();

    public double MontoExento { get; set; }
    public double MontoTotal { get; set; }

    public InformacionReferencia InformacionReferencia => _infoRef;

    protected override Dictionary<string, object?> BuildIdDoc()
    {
        var d = base.BuildIdDoc();
        d["FechaVencimientoSecuencia"] = Trim(FechaVencimientoSecuencia);
        return d;
    }

    protected override Dictionary<string, object?> BuildTotales() => new()
    {
        ["MontoExento"] = MontoExento.ToString("F2"),
        ["MontoTotal"] = MontoTotal.ToString("F2")
    };

    protected override InformacionReferencia? BuildInformacionReferencia() => _infoRef;
}

File: python-driver/src/chalona_driver/comprobantes/tipos_31_34.py
"""Comprobantes tipo 31 (FCF), 32 (Consumo), 33 (Nota Débito), 34 (Nota Crédito)."""

from .base import ComprobanteBase
from .datos import InformacionReferencia


class FacturaCreditoFiscal31(ComprobanteBase):
    """Factura de Crédito Fiscal (tipo 31)."""

    tipo_ecf = 31

    def __init__(self) -> None:
        super().__init__()
        self.monto_gravado_total = 0.0
        self.monto_gravado_i1 = 0.0
        self.itbis1 = 18
        self.total_itbis = 0.0
        self.total_itbis1 = 0.0
        self.monto_total = 0.0

    def _build_id_doc(self) -> dict:
        d = super()._build_id_doc()
        d["FechaVencimientoSecuencia"] = (self.fecha_vencimiento_secuencia or "").strip()
        d["IndicadorMontoGravado"] = str(self.indicador_monto_gravado)
        return d

    def _build_totales(self) -> dict:
        return {
            "MontoGravadoTotal": f"{self.monto_gravado_total:.2f}",
            "MontoGravadoI1": f"{self.monto_gravado_i1:.2f}",
            "ITBIS1": str(self.itbis1),
            "TotalITBIS": f"{self.total_itbis:.2f}",
            "TotalITBIS1": f"{self.total_itbis1:.2f}",
            "MontoTotal": f"{self.monto_total:.2f}",
        }


class FacturaConsumo32(ComprobanteBase):
    """Factura de Consumo (tipo 32). RFCE si MontoTotal < 250_000."""

    tipo_ecf = 32

    def __init__(self) -> None:
        super().__init__()
        self.monto_gravado_total = 0.0
        self.monto_gravado_i1 = 0.0
        self.itbis1 = 18
        self.total_itbis = 0.0
        self.total_itbis1 = 0.0
        self.monto_total = 0.0

    @property
    def monto_total_val(self) -> float:
        return self.monto_total

    def _build_id_doc(self) -> dict:
        d = super()._build_id_doc()
        if self.monto_total >= 250_000:
            d["FechaVencimientoSecuencia"] = (self.fecha_vencimiento_secuencia or "").strip()
        d["IndicadorMontoGravado"] = str(self.indicador_monto_gravado)
        return d

    def _build_totales(self) -> dict:
        return {
            "MontoGravadoTotal": f"{self.monto_gravado_total:.2f}",
            "MontoGravadoI1": f"{self.monto_gravado_i1:.2f}",
            "ITBIS1": str(self.itbis1),
            "TotalITBIS": f"{self.total_itbis:.2f}",
            "TotalITBIS1": f"{self.total_itbis1:.2f}",
            "MontoTotal": f"{self.monto_total:.2f}",
        }


class NotaDebito33(ComprobanteBase):
    """Nota de Débito (tipo 33). Requiere InformacionReferencia."""

    tipo_ecf = 33

    def __init__(self) -> None:
        super().__init__()
        self.monto_exento = 0.0
        self.monto_total = 0.0
        self._info_ref = InformacionReferencia()

    def _build_id_doc(self) -> dict:
        d = super()._build_id_doc()
        d["FechaVencimientoSecuencia"] = (self.fecha_vencimiento_secuencia or "").strip()
        return d

    def _build_totales(self) -> dict:
        return {
            "MontoExento": f"{self.monto_exento:.2f}",
            "MontoTotal": f"{self.monto_total:.2f}",
        }

    def _informacion_referencia(self) -> InformacionReferencia:
        return self._info_ref


class NotaCredito34(ComprobanteBase):
    """Nota de Crédito (tipo 34). Requiere InformacionReferencia."""

    tipo_ecf = 34

    def __init__(self) -> None:
        super().__init__()
        self.monto_exento = 0.0
        self.monto_total = 0.0
        self._info_ref = InformacionReferencia()

    def _build_id_doc(self) -> dict:
        d = super()._build_id_doc()
        d["FechaVencimientoSecuencia"] = (self.fecha_vencimiento_secuencia or "").strip()
        return d

    def _build_totales(self) -> dict:
        return {
            "MontoExento": f"{self.monto_exento:.2f}",
            "MontoTotal": f"{self.monto_total:.2f}",
        }

    def _informacion_referencia(self) -> InformacionReferencia:
        return self._info_ref
